// dvbjs/dist/src/lines.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const axios_1 = __importDefault(require("axios"));
const utils = __importStar(require("./utils"));
function parseDirection(direction) {
    return direction.Name;
}
function parseLine(line) {
    return {
        name: line.Name,
        mode: utils.parseMode(line.Mot),
        diva: utils.parseDiva(line.Diva),
        directions: line.Directions.map(parseDirection),
    };
}
/**
 * get a list of availible tram/bus lines for a stop.
 * @param stopID the stop ID
 * @param timeout the timeout of the request
 */
function lines(stopID, timeout = 5000) {
    const options = {
        url: "https://webapi.vvo-online.de/stt/lines",
        params: {
            format: "json",
            stopid: stopID,
        },
        timeout,
    };
    return axios_1.default(options)
        .then((response) => {
        // check status of response
        utils.checkStatus(response.data);
        let result = [];
        if (response.data.Lines) {
            result = response.data.Lines.map((l) => parseLine(l));
        }
        return result;
    })
        .catch(utils.convertError);
}
exports.lines = lines;
